cloud/amazon/cloudfront.py

```diff
@@ -0,0 +1,557 @@
+        params['DistributionId'] = module.params.get('resource_id')
+
+    params['Id'] = temp_results['Id']
+    resource_type = invocations[module.params.get('type')]
+    resource_details = json.loads(json.dumps(resource_type['get_resource_details'](**params), default=date_handler))
+
+    return resource_details
+
+
+def creation(client, module, policy):
+    changed = False
+    params = dict()
+    args = dict()
+
+    invocations = {
+        'distribution': {
+            'method': client.create_distribution,
+            'config_param': "DistributionConfig",
+            'result_key': "Distribution",
+        },
+        'origin_access_id': {
+            'method': client.create_cloud_front_origin_access_identity,
+            'config_param': "CloudFrontOriginAccessIdentityConfig",
+            'result_key': "CloudFrontOriginAccessIdentity",
+        },
+        'invalidation': {
+            'method': client.create_invalidation,
+            'config_param': "InvalidationBatch",
+            'result_key': "Invalidation",
+        },
+        'streaming': {
+            'method': client.create_streaming_distribution,
+            'config_param': "StreamingDistributionConfig",
+            'result_key': "StreamingDistribution",
+        },
+    }
+    resource_type = invocations[module.params.get('type')]
+    params[resource_type['config_param']] = policy
+    if module.params.get('type') == "invalidation":
+        params['DistributionId'] = module.params.get('resource_id')
+        args['DistributionId'] = module.params.get('resource_id')
+
+    invocation = resource_type['method']
+    result = json.loads(json.dumps(invocation(**params), default=date_handler))
+
+    if module.params.get('wait_for_deployed'):
+        args['Id'] = result[resource_type['result_key']]['Id']
+        status_achieved = wait_for_deployed_status(client, module, **args)
+        if not status_achieved:
+            module.fail_json(msg="Timed out waiting for the resource to finish"
+                                " deploying, please check the AWS console for"
+                                " the latest status.")
+
+    changed = True
+    return (changed, result)
+
+
+def update(client, module, policy, resource_id, etag):
+    changed = False
+    params = dict()
+    invocations = {
+        'distribution': {
+            'method': client.update_distribution,
+            'config_param': "DistributionConfig"
+            },
+        'origin_access_id': {
+            'method': client.update_cloud_front_origin_access_identity,
+            'config_param': "CloudFrontOriginAccessIdentityConfig"
+            },
+        'streaming': {
+            'method': client.create_streaming_distribution,
+            'config_param': "StreamingDistributionConfig",
+            },
+    }
+
+    params[invocations[module.params.get('type')]['config_param']] = policy
+    params['Id'] = resource_id
+    params['IfMatch'] = etag
```

Comment:
Hi @adamchainz, the etag is actually grabbed from the existing CDN itself - in creation_setup I only call update when there is an existing CDN, I then grab the etag from the lookup and pass it to the update function.